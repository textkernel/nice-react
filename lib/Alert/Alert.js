import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import niceProps from '../props';
import './Alert.scss';

class Alert extends Component {
    renderHeading() {
        const { heading } = this.props;
        if (!heading) {
            return null;
        }
        return (
            <div className="nice-alert-heading">
                { heading }
            </div>
        );
    }

    renderButton() {
        const { buttonLabel = '', onClick } = this.props;
        if (!onClick) {
            return null;
        }
        if (!buttonLabel.length) {
            return (
                <a href="#" className="nice-alert-dismiss" onClick={ e => {
                    e.preventDefault();
                    onClick();
                }}>
                    <i className="icon fa-angle-down" />
                </a>
            );
        }
        return (
            <a href="#" className="nice-alert-button" onClick={ e => {
                e.preventDefault();
                onClick();
            }}>
                { buttonLabel }
            </a>
        );
    }

    render() {
        const { buttonLabel, children, context, className, onClick, position, style } = this.props;
        let classes = ['nice-alert', context];
        const alertPositions = new Set(niceProps.alertPositions);

        if (position) {
            const intersection = position.split(' ').filter(p => alertPositions.has(p)) || [];

            if (intersection.length) {
                classes.push('position-fixed');
                classes = classes.concat(intersection);
            }
        }

        if (onClick && buttonLabel) {
            return (
                <div
                    className={ classNames(classes, className) }
                    style={ style }
                >
                    <div className="nice-alert-wrapper">
                        <div className="nice-alert-content">
                            { children }
                        </div>
                        { this.renderButton() }
                    </div>
                </div>
            );
        }

        return (
            <div
                className={ classNames(classes, className) }
                style={ style }
            >
                { this.renderHeading() }
                { children }
                { this.renderButton() }
            </div>
        );
    }
}

Alert.propTypes = {
    context: PropTypes.oneOf(niceProps.alertContexts),
    className: PropTypes.string,
    heading: PropTypes.string,
    children: PropTypes.node,
    buttonLabel: PropTypes.string,
    onClick: PropTypes.func,
    position: PropTypes.string,
    style: PropTypes.object
};

Alert.defaultProps = {
    context: 'info',
    className: '',
    heading: null,
    children: null,
    buttonLabel: '',
    onClick: null,
    position: null,
    style: null
};

export default Alert;
