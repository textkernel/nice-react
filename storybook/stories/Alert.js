import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobsOptions, withKnobs, text, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import props from '../../lib/props';

import { Alert } from 'nice-react';

const stories = storiesOf('Alert', module);

stories.addDecorator((story, context) => withInfo('')(story)(context));
stories.addDecorator(withKnobs);

stories.add('Default', () => (
    <Alert>
    	<strong>Hey you!</strong> Here is some information
    </Alert>
));

stories.add('With heading', () => (
    <Alert heading="Warning" context="warning">
    	<strong>Heads up</strong> This is a warning
    </Alert>
));

stories.add('Dismissable', () => (
    <Alert onClick={ () => { alert('dismiss'); }} context="bad">
    	<strong>Uh oh...</strong> Something is wrong
    </Alert>
));

stories.add('With button', () => (
    <Alert onClick={ () => { alert('dismiss'); }} context="good" buttonLabel="Dismiss">
    	<strong>Great success</strong> All good!
    </Alert>
));

stories.add('Custom properties', () => {
    const contextOptions = props.alertContexts;
    const context = select('Context', contextOptions, props.alertContexts[0]);
    const message = text('Message', 'Change me');
    const heading = text('Heading', '');
    const buttonLabel = text('Button label', '');

    return <Alert context={ context } heading={ heading } buttonLabel={ buttonLabel } onClick={ () => {} }>
    	{ message }
    </Alert>
});
