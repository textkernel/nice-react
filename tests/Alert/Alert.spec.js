import React from 'react';
import Alert from '../../lib/Alert';

describe('<Alert />', () => {
    it('should render the default Alert', () => {
        const wrapper = shallow(<Alert>
        <strong>Hey you!</strong> Here is some information
    </Alert>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render Alert with heading', () => {
        const wrapper = shallow(<Alert heading="Warning" context="warning">
            <strong>Heads up</strong> This is a warning
        </Alert>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render Alert with icon button', () => {
        const onClickHandler = jest.fn();
        const wrapper = shallow(<Alert onClick={ onClickHandler } context="bad">
            <strong>Uh oh...</strong> Something is wrong
        </Alert>);
        expect(wrapper).toMatchSnapshot();

        wrapper.find('.nice-alert-dismiss').simulate('click', { preventDefault() {} });
        expect(onClickHandler).toHaveBeenCalled();
    });

    it('should render Alert with large button', () => {
        const onClickHandler = jest.fn();
        const wrapper = shallow(<Alert onClick={ onClickHandler } context="good" buttonLabel="Dismiss">
            <strong>Great success</strong> All good!
        </Alert>);
        expect(wrapper).toMatchSnapshot();

        wrapper.find('.nice-alert-button').simulate('click', { preventDefault() {} });
        expect(onClickHandler).toHaveBeenCalled();
    });
});
