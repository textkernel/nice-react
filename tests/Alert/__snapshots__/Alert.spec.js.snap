// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should render Alert with custom classes / styles 1`] = `
<div
  className="nice-alert info test"
  style={
    Object {
      "display": "none",
    }
  }
>
  <strong>
    Hey you!
  </strong>
   Here is some information
</div>
`;

exports[`<Alert /> should render Alert with heading 1`] = `
<div
  className="nice-alert warning"
  style={null}
>
  <div
    className="nice-alert-heading"
  >
    Warning
  </div>
  <strong>
    Heads up
  </strong>
   This is a warning
</div>
`;

exports[`<Alert /> should render Alert with icon button 1`] = `
<div
  className="nice-alert bad"
  style={null}
>
  <strong>
    Uh oh...
  </strong>
   Something is wrong
  <a
    className="nice-alert-dismiss"
    href="#"
    onClick={[Function]}
  >
    <i
      className="icon fa-angle-down"
    />
  </a>
</div>
`;

exports[`<Alert /> should render Alert with large button 1`] = `
<div
  className="nice-alert good"
  style={null}
>
  <div
    className="nice-alert-wrapper"
  >
    <div
      className="nice-alert-content"
    >
      <strong>
        Great success
      </strong>
       All good!
    </div>
    <a
      className="nice-alert-button"
      href="#"
      onClick={[Function]}
    >
      Dismiss
    </a>
  </div>
</div>
`;

exports[`<Alert /> should render the default Alert 1`] = `
<div
  className="nice-alert info"
  style={null}
>
  <strong>
    Hey you!
  </strong>
   Here is some information
</div>
`;
